rust2vec/src/storage.rs

```diff
@@ -47,39 +49,127 @@ pub type CowArray1<'a, A> = CowArray<'a, A, Ix1>;
 /// To allow for embeddings to be stored in different manners (e.g.
 /// regular *n x d* matrix or as quantized vectors), this trait
 /// abstracts over concrete storage types.
-#[derive(Debug, PartialEq)]
+#[derive(Debug)]
 pub enum Storage {
+    /// In-memory `ndarray` matrix.
     NdArray(Array2<f32>),
+
+    /// Memory-mapped matrix.
+    Mmap { map: Mmap, shape: Ix2 },
 }
 
 impl Storage {
     pub(crate) fn chunk_identifier(&self) -> ChunkIdentifier {
         match self {
             Storage::NdArray(_) => ChunkIdentifier::NdArray,
+            Storage::Mmap { .. } => ChunkIdentifier::NdArray,
         }
     }
 
     pub fn dims(&self) -> usize {
         match self {
             Storage::NdArray(ref data) => data.cols(),
+            Storage::Mmap { shape, .. } => shape[1],
         }
     }
 
     pub fn embedding(&self, idx: usize) -> CowArray1<f32> {
-        match self {
-            Storage::NdArray(ref data) => CowArray::Borrowed(data.index_axis(Axis(0), idx)),
-        }
+        CowArray::Owned(self.view().row(idx).to_owned())
     }
 
     pub fn view(&self) -> ArrayView2<f32> {
         match self {
             Storage::NdArray(ref data) => data.view(),
+            Storage::Mmap { map, shape } => unsafe {
+                ArrayView2::from_shape_ptr(*shape, map.as_ptr() as *const f32)
+            },
+        }
+    }
+
+    fn write_ndarray_chunk<S, W>(data: ArrayBase<S, Ix2>, mut write: W) -> Result<(), Error>
+    where
+        S: Data<Elem = f32>,
+        W: Write + Seek,
+    {
+        // n_rows: 8 bytes, n_cols: 4 bytes, type_id: 4, matrix
+        let chunk_len = 16 + (data.rows() * data.cols() * mem::size_of::<f32>());
+
+        write.write_u32::<LittleEndian>(ChunkIdentifier::NdArray as u32)?;
+        write.write_u64::<LittleEndian>(chunk_len as u64)?;
+        write.write_u64::<LittleEndian>(data.rows() as u64)?;
+        write.write_u32::<LittleEndian>(data.cols() as u32)?;
+        write.write_u32::<LittleEndian>(f32::type_id())?;
+
+        // Write padding, such that the embedding matrix starts on at
+        // a multiple of the size of f32 (4 bytes). This is necessary
+        // for memory mapping a matrix. Interpreting the raw u8 data
+        // as a proper f32 array requires that the data is aligned in
+        // memory. However, we cannot always memory map the starting
+        // offset of the matrix directly, since mmap(2) requires a
+        // file offset that is page-aligned. Since the page size is
+        // always a larger power of 2 (e.g. 2^12), which is divisable
+        // by 4, the offset of the matrix with regards to the page
+        // boundary is also a multiple of 4.
+
+        let n_padding = padding::<f32>(write.seek(SeekFrom::Current(0))?);
+        let padding = vec![0; n_padding as usize];
+        write.write(&padding)?;
+
```

Comment:
This comment is to verify that I'm correctly reading this method:
`write.seek(SeekFrom::Current(0))?);` returns the current position of the writer and `padding::<f32>(position)` then returns the bytes missing to hit the alignment to multiples of `f32`.

Reply:
Correct on both accounts.